### PR TITLE
Replace `-m torch.distributed.run` by `torchrun`

### PR DESCRIPTION
**Pull request by regisss**
# What does this PR do?

This PR replaces occurrences of `-m torch.distributed.launch` (deprecated) and `-m torch.distributed.run` (equivalent) by `torchrun`. More information [here](https://pytorch.org/docs/stable/elastic/run.html).

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ArthurZucker** [comment]
Just for reference, is there a reason why the previous occurence is deprecated? (not familiar with it!)

**Comment by regisss** [comment]
`torchrun` is equivalent to `python -m torch.distributed.run` while `python -m torch.distributed.launch` is deprecated. I think the reason why it is deprecated is just that `torchrun` does the same but also provides more functionalities.

I improved the description of this PR accordingly.

**Comment by regisss** [comment]
However `torchrun` has only been available since the release of `torch` 1.10. I guess we want to keep compatibility with some previous versions of `torch` right? @sgugger @ArthurZucker

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1684,8 +1684,7 @@ def _inner_training_loop(
                 # nn.DataParallel(model) replicates the model, creating new variables and module
                 # references registered here no longer work on other gpus, breaking the module
                 raise ValueError(
-                    "Currently --debug underflow_overflow is not supported under DP. Please use DDP"
-                    " (torch.distributed.launch)."
+                    "Currently --debug underflow_overflow is not supported under DP. Please use DDP" " (torchrun)."
```
Comment (sgugger) [inline src/transformers/trainer.py:1687]:
```suggestion
                    "Currently --debug underflow_overflow is not supported under DP. Please use DDP (torchrun)."
```

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -1503,7 +1503,7 @@ def _setup_devices(self) -> "torch.device":
         if torch.distributed.is_available() and torch.distributed.is_initialized() and self.local_rank == -1:
             logger.warning(
                 "torch.distributed process group is initialized, but local_rank == -1. "
-                "In order to use Torch DDP, launch your script with `python -m torch.distributed.launch"
+                "In order to use Torch DDP, launch your script with `torchrun"
```
Comment (sgugger) [inline src/transformers/training_args.py:1506]:
```suggestion
                "In order to use Torch DDP, launch your script with `torchrun`"
```

File: tests/extended/test_trainer_ext.py [hunk 3]
```diff
@@ -366,12 +366,11 @@ def run_trainer(
                 n_gpus_to_use = get_gpu_count()
             master_port = get_torch_dist_unique_port()
             distributed_args = f"""
-                -m torch.distributed.launch
                 --nproc_per_node={n_gpus_to_use}
                 --master_port={master_port}
                 {self.examples_dir_str}/pytorch/translation/run_translation.py
             """.split()
-            cmd = [sys.executable] + distributed_args + args
+            cmd = ["torchrun"] + distributed_args + args
```
Comment (sgugger) [inline tests/extended/test_trainer_ext.py:373]: If this one is changed, the test will need to be decorated to require a more recent version of PyTorch.

File: tests/trainer/test_trainer_distributed.py [hunk 4]
```diff
@@ -67,14 +66,13 @@ class TestTrainerDistributedNeuronCore(TestCasePlus):
...
         distributed_args = f"""
-            -m torch.distributed.launch
             --nproc_per_node=2
             --master_port={get_torch_dist_unique_port()}
             {self.test_file_dir}/test_trainer_distributed.py
         """.split()
         output_dir = self.get_auto_remove_tmp_dir()
         args = f"--output_dir {output_dir}".split()
-        cmd = [sys.executable] + distributed_args + args
+        cmd = ["torchrun"] + distributed_args + args
```
Comment (sgugger) [inline tests/trainer/test_trainer_distributed.py:75]: Same in this file.

**Review by sgugger — COMMENTED** [review]
Yes we would need to add some notes for earlier PyTorch version as we officially support 1.7 to 1.13. Mostly in the PyTorch README and the run_scripts doc.

Also please focus the PR on the changes of commands only and remove all styling nits (you are welcome to suggest them in another PR) as it's distracting a lot the review.

**Pull request closed without merging**